My Jetpack: Fix Current_Plan::get() not getting current/latest plan. Cache stuck.

elliottprogrammer opened the pull request:
This PR fixes an issue in My Jetpack that is causing incorrect results in `[Product class]::get_paid_bundles_that_include_product()` which in turn produces the wrong result in `[Product class]:get_paid_purchase_for_product()`, which ultimately is affecting the red-bubble `alert_if_paid_plan_requires_plugin_install_or_activation()` causing the red-bubble notice/banner to report the incorrect plugins needing installed or activated (because `Current_Plan::get()` is stuck cached and always returning a stale result (See also: p1738167022814599/1738164593.019149-slack-C07RQ9B3UFK).

This bug/issue was first introduced in a bug-fix PR: [Package protect-status: Fix Current_Plan::supports from busting cache on every call](https://github.com/Automattic/jetpack/pull/41010), which prevents `Current_Plan::supports()` from pulling new wpcom api data on every call, basically allowing the result to be cached.  Issue discussion here: p1736794507070729-slack-C02LK1W8T4Z
The problem is, now `Current_Plan::supports()` is now over-aggressively cached and only always returning stale results (in My Jetpack) after purchasing (or removing) a new paid plan/bundle.

<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* In My Jetpack, update `static::get_site_current_plan(...)` to refresh the data from wpcom. 

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

It would be best to test this PR with your local Jetpack dev environment (via Docker), as opposed to Jurassic Ninja

1. Running the latest `trunk` branch (Bleeding edge).
2. Make sure only the VideoPress plugin is active. (Jetpack plugin **not** active, only VideoPress active).
3. Make sure Jetpack is connected (both site and user).
4. First make sure your site has no paid plans (only Jetpack Free).
5. From Jetpack root, run `jetpack docker wp option get jetpack_active_plan`
6. If the returned active plan is Not Jetpack Free, do the following steps:
    - Checkout this branch: `git checkout fix/mj-current-plan-not-updated`
    - Load/open the My Jetpack page.
    - Now checkout `trunk` branch again.
    - Run `jetpack docker wp option get jetpack_active_plan`, confirm Jetpack Free is the active plan
7. Purchase Jetpack Complete plan.
8. Open My Jetpack.
9. See/confirm that the plugin activation/installation notice is showing, and that it's saying only "CRM" needs activated. See screenshot:
10. ![Screen Shot 2025-01-30 at 13 38 00](https://github.com/user-attachments/assets/84163d28-21c3-42ff-9ec5-cc5071d1b4b4)
11. You can run `jetpack docker wp option get jetpack_active_plan` again, and confirm that the active plan is still Jetpack Free, even though you purchased Jetpack Complete. This is the bug. The `Current_Plan::get()` is stuck with stale data.
12. Now checkout this PR branch: `git checkout fix/mj-current-plan-not-updated`
13. Reload My Jetpack page.
14. See that the plugin activation/installation notice is now showing all the Jetpack plugins needing activated. See screenshot:
15. ![Screen Shot 2025-01-30 at 13 47 25](https://github.com/user-attachments/assets/7bba7413-7ff1-4a85-bc92-f30e54e317c4)
16. You can run `jetpack docker wp option get jetpack_active_plan` again, to confirm that the active plan is now refreshed to the correct plan, "Jetpack Complete".
17. Activate the "Jetpack" plugin. You should now see that only Akismet, Boost, and CRM are the required plugins needing activated for Jetpack Complete to function fully. See screenshot:
18. ![Screen Shot on 2025-01-30 at 13-57-26](https://github.com/user-attachments/assets/da03d019-c171-4db3-9c36-2bea7c3ec6af)

elliottprogrammer committed:
Fix Current_Plan::get() not getting current/latest plan. Cache stuck.

elliottprogrammer committed:
changelog

jboland88 commented on projects/packages/my-jetpack/src/products/class-product.php:
```diff
@@ -435,7 +435,7 @@ public static function get_standalone_info() {
 		// First check site features (if there's a feature that identifies the paid plan)
 		if ( static::$feature_identifying_paid_plan ) {
-			if ( static::does_site_have_feature( static::$feature_identifying_paid_plan ) ) {
+			if ( Wpcom_Products::current_plan_supports( static::$feature_identifying_paid_plan ) ) {
```
My initial thought is that this might break some things unintentionally since the feature check is meant to take into account all purchases on a site and `current_plan_supports` is only going to check against the current plan (not add on purchases).

I am wondering if there's a different place we can fix the issue 🤔

elliottprogrammer replied:
Oh hmm, I may be mistaken, but I think `Current_Plan::supports( $feature )` checks agains all the site's owned products too.  At least that was my understanding when I did this.  Let me look closer into it to confirm..

elliottprogrammer replied:
@jboland88, Yeah, I believe `static::does_site_have_feature( static::$feature_identifying_paid_plan )` and `Wpcom_Products::current_plan_supports( static::$feature_identifying_paid_plan )` are both comparing the `$feature_identifying_paid_plan` to the same set of site features (the sites 'active' features).

The previous way ( static::does_site_have_feature() ) pulls the sites 'active' features from `/rest/v1.1/sites/229710838/features` and from the response data `active` property.

The Current_Plan way pulls the sites 'active' features from `/rest/v1.1/sites/229710838` and from the response data: `plan->features->active` property.

But they're basically both doing the same thing (comparing the same things), to my understanding.

elliottprogrammer replied:
Ok, after a Slack discussion: p1738263782787889-slack-C07RQ9B3UFK, we're going to take a slightly different approach:  Like this:

```php
/**
 * Gets the site's currently active "plan" (bundle).
 *
 * @return array
 */
public static function get_site_current_plan( $reload = false ) {
	static $reloaded_already = false;

	if ( $reload && ! $reloaded_already ) {
		Current_Plan::refresh_from_wpcom();
		$reloaded_already = true;
	}

	return Current_Plan::get();
}
```

elliottprogrammer committed:
Refactor: Update 'static::does_site_have_feature(...)' instead.

elliottprogrammer commented on projects/packages/my-jetpack/_inc/hooks/use-notification-watcher/use-paid-plan-needs-plugin-install-activation-notice.tsx:
```diff
@@ -57,13 +57,14 @@ const usePaidPlanNeedsPluginInstallActivationNotice = ( redBubbleAlerts: RedBubb
 
 	const getPluginInfo = useCallback(
-		productSlug => ( {
+		( productSlug, action ) => ( {
```
The changes in this file are unrelated to the cacheing fix for the Current_Plan class. It is just a last-minute improvement to allow each plugin (in the displayed list of plugins needing activated or installed) to display individually whether it needs "installed and activated" or only "activated". Prior to this change, they all displayed the same thing (the same action noun).

elliottprogrammer commented:
@jboland88, Thanks for reviewing and for your feedback! I've refactored `static::get_site_current_plan( $reload = false )` as we discussed: p1738271946481979/1738263782.787889-slack-C07RQ9B3UFK (and the [comments above](https://github.com/Automattic/jetpack/pull/41421#discussion_r1936180243))
Would you mind taking another look?  Thanks! 🙏😊

jboland88 commented on projects/packages/my-jetpack/_inc/hooks/use-notification-watcher/use-paid-plan-needs-plugin-install-activation-notice.tsx:
```diff
@@ -184,18 +192,30 @@ const usePaidPlanNeedsPluginInstallActivationNotice = ( redBubbleAlerts: RedBubb
-						) ) }
+						{ pluginsList.map( ( pluginInfo, index ) => {
+							const pluginActionNoun =
+								pluginInfo.action === 'activate' ? 'activated' : 'installed and activated';
```
I think this needs to read "activation" or "installation and activation" so that it makes sense when added with the "needs string".

I'd also advise making a separate function that translates the two full strings separately, I don't think one string can be matched with multiple variable outputs in the translation system.

IanRamosC replied:
Following the translation guidelines, we should have that as separate strings, as Josh mentioned above. Some strings might lose their meaning when translated separately. I think we can create a quick mapping of actions -> translations.

jboland88 commented on projects/packages/my-jetpack/src/products/class-product.php:
```diff
@@ -514,7 +514,7 @@ public static function get_paid_bundles_that_include_product() {
 		}
 		$paid_bundles   = $features['available']->$idendifying_feature ?? array();
-		$current_bundle = Wpcom_Products::get_site_current_plan();
+		$current_bundle = Wpcom_Products::get_site_current_plan( true );
```
The only other suggestion I'd have here would be adding a transient cache for about 5 minutes so that we are not re-fetching the plan every time My Jetpack loads.

This is not a blocker for this PR.

Even better would be to cache the result for a long time and invalidate the cache when we detect a change in the site's purchase data - which we already fetch for My Jetpack. But that's better handled in a follow-up task.

IanRamosC replied:
Personally, I'd avoid merging this PR without at least some level of caching, just to avoid what Josh said about re-fetching every time. I don't oppose to being in a different PR as long as this is included in the same release.

jboland88 reviewed (changes requested):
One translation issue that needs to be fixed, but otherwise this looks OK

IMO it would be better to split off the changes in `use-paid-plan-needs-plugin-install-activation-notice.tsx` to a separate PR and handle the change there.

There's some opportunity to improve caching on the current plan fetch so we don't reload that with every visit to My Jetpack, but that can be considered in another PR

IanRamosC reviewed (commented):
The change looks good in general, I just left some minor comments similar to what Josh mentioned earlier.

elliottprogrammer committed:
Add transient cache & remove notice translation changes.

elliottprogrammer committed:
get_site_current_plan( true ): remove unnecessary true argument.

elliottprogrammer commented:
Thanks @jboland88 & @IanRamosC , 🙏
I've added the transient cacheing and removed the text/translation changes in the notice file. I'll re-update those changes in a new PR. 👍 How's it look now? 😊

elliottprogrammer committed:
Set transient to 5 min (360 sec).

elliottprogrammer committed:
Set transient to 5 min (360 sec).

elliottprogrammer committed:
Revert back to 'get_site_current_plan(..) without transient.

elliottprogrammer committed:
Pass 'true' to static::get_site_current_plan(...), to refresh data.

jboland88 reviewed (approved):
This works and the code looks good.

I tested with Query Monitor just to confirm that the /sites endpoint used to get the current plan was not being called outside of My Jetpack 👍 

As mentioned in other discussions, we'll work on a follow-up PR to allow us to cache the current plan and invalidate it when new purchases are detected for the site.
